### ISSUE TITLE
[3.0.0]Seg fault on php5 when forwarding from beforeException

### ISSUE THREAD

**Issue by Jurigag** [phalcon/cphalcon#12154]
### Expected and Actual Behavior

Should return test, but it doesn't, there is just seg fault when accessing just index page(`/`)

**On php7 it works though.**

Backtrace - http://pastebin.com/jD5GQJaJ

Github repo - https://github.com/Jurigag/phalcon-segfault-before-forward
### Details
- Phalcon version:

```
Web framework delivered as a C-extension for PHP
phalcon => enabled
Author => Phalcon Team and contributors
Version => 3.0.1
Build Date => Aug 20 2016 15:06:39
Powered by Zephir => Version 0.9.4a-dev-121e9b4bf1

Directive => Local Value => Master Value
phalcon.db.escape_identifiers => On => On
phalcon.db.force_casting => Off => Off
phalcon.orm.events => On => On
phalcon.orm.virtual_foreign_keys => On => On
phalcon.orm.column_renaming => On => On
phalcon.orm.not_null_validations => On => On
phalcon.orm.exception_on_failed_save => Off => Off
phalcon.orm.enable_literals => On => On
phalcon.orm.late_state_binding => Off => Off
phalcon.orm.enable_implicit_joins => On => On
phalcon.orm.cast_on_hydrate => Off => Off
phalcon.orm.ignore_unknown_columns => Off => Off
```
- PHP Version:

```
PHP 5.6.19 (cli) (built: Apr 24 2016 12:31:51) 
Copyright (c) 1997-2016 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2016 Zend Technologies
    with Zend OPcache v7.0.6-dev, Copyright (c) 1999-2016, by Zend Technologies
```
- Operating System: Ubuntu 14.04
- Installation type: zephir
- Zephir version (if any): Version 0.9.4a-dev-121e9b4bf1
- Server: Apache

**Comment by virgofx** [comment]
Likely related to: #11819  -- Can we mark both of these as relevant for 3.0.2 milestone as this has been an issue for awhile.

**Comment by Jurigag** [comment]
Yea but what is weird it works on php 7 though.

**Comment by virgofx** [comment]
Confirmed issue as well.

*Backtrace*
http://vpaste.net/wGczw

**Comment by virgofx** [comment]
There are actually 2 bugs. 

1. A return `false` when forwarding dispatcher in exception will cause sigfault
2. Exception handler is called twice when the dispatcher is not marked as finished

Here is the test script which outlines the 2 bugs.

```php
<?php

/**
 * IndexController
 */
class IndexController extends \Phalcon\Mvc\Controller
{
    public function initialize()
    {
        $this->getDI()->getShared('view')->disable();
    }

    public function indexAction()
    {
        echo 'Index Action <br />';

        throw new Exception('Something happened');
    }

    public function errorAction()
    {
        echo 'Error Action <br />';
    }
}

/**
 * Dispatcher ExceptionHandler
 */
class ExceptionHandler extends \Phalcon\Mvc\User\Plugin
{
    public function beforeException(\Phalcon\Events\Event $event, \Phalcon\DispatcherInterface $dispatcher, \Exception $exception)
    {
        echo 'ExceptionHandler::beforeException() - ' . $exception->getMessage() . '<br />';


        ////////////////////////////////////////////////////////
        // Case 1:
        ////////////////////////////////////////////////////////

        // return false;

        //**Output** (OK)
        //      Index Action
        //      ExceptionHandler::beforeException() - Something happened


        ////////////////////////////////////////////////////////
        // Case 2 + 3 - Empty return or nothing at all
        ////////////////////////////////////////////////////////

        // return;

        //**Output** (Exception handler gets thrown twice)
        //      Index Action
        //      ExceptionHandler::beforeException() - Something happened
        //      ExceptionHandler::beforeException() - Something happened
        //      Fatal error: Uncaught exception 'Exception' with message 'Something happened' in /var/www/public/index.php:18 Stack trace: #0 [internal function]: IndexController->indexAction() #1 [internal function]: Phalcon\Dispatcher->callActionMethod(Object(IndexController), 'indexAction', Array) #2 [internal function]: Phalcon\Dispatcher->_dispatch() #3 [internal function]: Phalcon\Dispatcher->dispatch() #4 /var/www/public/index.php(87): Phalcon\Mvc\Application->handle() #5 {main} thrown in /var/www/public/index.php on line 18

        // The beforeException gets called again (resulting 2x) because we're not marked as finished so the exception gets
        // thrown up the stack
        // https://github.com/phalcon/cphalcon/blob/master/phalcon/dispatcher.zep#L595


        ////////////////////////////////////////////////////////
        // Case 4:
        ////////////////////////////////////////////////////////

        // $dispatcher->forward(['action' => 'error']);
        // return false;

        //**Output** (SIGFAULT - Combination of forward + return false)
    }
}


////////////////////////////////////////////////////////
// Bootstrap
////////////////////////////////////////////////////////

$di = new \Phalcon\Di\FactoryDefault();

$di->setShared('view', function() {
    return new \Phalcon\Mvc\View();
});

$di->setShared('dispatcher', function() {
    $eventsManager = new \Phalcon\Events\Manager();
    $eventsManager->attach('dispatch:beforeException', new ExceptionHandler());
    $dispatcher = new \Phalcon\Mvc\Dispatcher();
    $dispatcher->setEventsManager($eventsManager);
    return $dispatcher;
});

$application = new \Phalcon\Mvc\Application();
$application->setEventsManager(new \Phalcon\Events\Manager());
$application->setDI($di);

echo $application->handle()->getContent();

```

**Comment by virgofx** [comment]
Prior to following commit works - 

https://github.com/phalcon/cphalcon/commit/3fcfff0fd6a53fcb5bef52a4d003ca5515bdf380#diff-074ce759b4ecc1363a895a6a47faf187

**Comment by sergeyklay** [comment]
@virgofx

Could you please check 3.0.x branch?

```sh
giit clone git@github.com:phalcon/cphalcon.git
git checkout 3.0.x

# Use latest Zephir here from master branch
zephir fullclean
zephir build
```

**Comment by sergeyklay** [comment]
Fixed in the `3.2.x` branch.
